### DISCUSSION TITLE
Run actions after succesfully logged in

### DISCUSSION THREAD

**Discussion by crisfcodes** [react-navigation/react-navigation#10718] [Q&A]
I'm creating a React Native app in which some actions like adding to favorites require the user to be logged in.

## The problem

If a certain action needs authentication, the following flow is executed:

1. User tab over the favorite button(Protected action)
2. A modal(screen with presentation: "modal") is rendered to allow the user to enter their credentials
3. If the user is successfully logged in, the modal is closed and the user is directed to the screen on which it was located(`goBack()` navigation action).
4. **THE PROBLEM**: user needs to press again over the favorite button, the idea is, if the user is successfully logged in, the action (add to favorites) is executed immediately without the user having to perform the action again.

**NOTE**: I can have different protected actions on the same screen

## Question
how can I request the user to log in to perform the action and have the action executed automatically after successful login?
**execute the protected action only once, only when the user logs in successfully and the modal is closed, if the user is already authenticated the protected action should not be executed again.**

## Example flow

```typescript
function FavoriteScreen({ navigation }) {
  const { isAuthenticated } = useAuth();

  if (isAuthenticated) {
      addFavorite(productId);
  } else {
    navigation.navigate("LoginScreen");
  }
}
```

## Things I've tried

1. Send a callback through the parameters(called `nexAction`) from the protected action screen to the login screen, run the callback after successfully logging in, and close the modal, but I'm getting [non-serializable](https://reactnavigation.org/docs/troubleshooting/#i-get-the-warning-non-serializable-values-were-found-in-the-navigation-state) warning, and this screen implements deep linking, so I cannot ignore the warning as the documentation suggests.

```typescript
if (isAuthenticated) {
  addFavorite();
} else {
  navigation.navigate(NavigationScreens.LoginScreen, {
    nextAction: addFavorite,
  });
}
```

2. Use the `focus` event to run the protected action after the user is successfully logged in and the modal is closed. This approach has some problems, each time user focuses on the screen and is authenticated, the protected action is going to be run, and there may be more than one protected action on the screen, meaning that they will all be executed when the screen is focused. 

```typescript
useEffect(() => {
  const unsubscribe = navigation.addListener('focus', () => {
    if (isAuthenticated) {
      addFavorite();
    }
  });

  return unsubscribe;
}, [isAuthenticated, navigation, addFavorite]);
```

**Comment by satya164** [comment]
This is unrelated to navigation. You'd have some kind of listener which runs on login and do the action there. Rough code:

```js
  if (isAuthenticated) {
    addFavorite(productId);
  } else {
    onLogin(() => addFavorite(productId));
    navigation.navigate("LoginScreen");
  }
```

**Reply by crisfcodes** [reply]
Hey @satya164 can u expand ur answer is not clear to me what should be `onLogin`

**Reply by satya164** [reply]
A function which calls a callback after login - a very rough example, but it's not important what `onLogin` is. The point is that you want to do something on login, not on a screen being focused. So you shouldn't think this in terms of navigation or screens, think in terms of how to do something after login.
